WordPress/gutenberg
# Query Loop: Use templateSlug and postType for more context

## What?

The Query Loop block currently does not have much context about where it has been added when viewed in the Editor, and in some cases, it does not always make sense to inherit the global query. In https://github.com/WordPress/gutenberg/pull/65067, we discussed that these cases can mostly be split between whether the block has been added to singular content, such as a single post or page; or non-singular content, such as the archive template or category template.

We addressed some of this logic in https://github.com/WordPress/gutenberg/pull/65067, however, this only covered cases where the block is added to single posts or pages. As a follow-up to https://github.com/WordPress/gutenberg/pull/65067, this PR also considers the 404, blank, and custom templates as singular content while also refactoring this recently added logic. 

This PR does not add any other context for when the block is added to non-singular content, but I think we should address these cases in follow-up PRs, including:
- Specific taxonomy templates
- Date archive template
- Author archive and single author template

## Why?

Part of https://github.com/WordPress/gutenberg/issues/30369, and a follow-up to https://github.com/WordPress/gutenberg/pull/65067.

## How?

- Adds `templateSlug` to the context available to the Query Loop block, which allows us to check which template is in use before deciding whether to inherit the query. There is some prior art in https://github.com/WordPress/gutenberg/pull/27972.
- Moves the main logic to a function that can be tested more easily and is hopefully easier to extend.
- Applies the closest `postType`, if available, to the current query.

## Testing Instructions




The functionality isn't too different to trunk, other than the 404, blank and custom templates are now included in the list of "singular content" contexts (i.e. `isSingular`). The `postType` is also being applied where it's available, so please also test templates and content for custom post types.

1. In the Editor, create or edit templates for singular (single post, single page, 404, blank, custom) and non-singular (archive, authors, category, tag) content.
2. Insert a Query Loop block in each.
3. Ensure the block defaults to `inherit: false` in singular content.
4. Ensure the block defaults to `inherit: true` in non-singular content.
5. Ensure queries in a template or content for a custom post type automatically query the same post type.

## Screenshots or screencast 

| Query Loop when added to singular content | Query Loop when added to non-singular content |
| --------- | --------- |
| <img width="519" alt="image" src="https://github.com/user-attachments/assets/d168a670-bf3b-401e-a853-948c759c6062"> | <img width="410" alt="image" src="https://github.com/user-attachments/assets/b79f0bd6-b424-4676-abaf-f01dcdd085b4"> |
